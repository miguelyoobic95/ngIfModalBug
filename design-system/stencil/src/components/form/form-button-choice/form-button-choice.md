---
name: Button Choice
category: Form Components
---

```yoo-form-button-choice.html

    <yoo-form-button-choice></yoo-form-button-choice>

```
```yoo-form-button-choice.js

    var comp = document.querySelector('yoo-form-button-choice');
    comp.choices = ['salut', 'bonjour', 'hi'];

```

## Attributes

|Name|Type|Default|Description|
|---|---|---|---|
|`multiple`|boolean|`false`|Set to true if you want to allow multiple selection|
|`choices`|`string[]`|`[]`|Set of choices to print in the buttons|
|`selected`|`boolean[]`|`[]`|Pre defined selected items. Use this prop to restaure a previous state. Set true for the item which need to be selected at the beginning|

If you try to select more than one choice without setting `multiple` to true, choices will be all unselected. 
### Multiple Attribute

```yoo-form-button-choice-multiple.html

    <yoo-form-button-choice multiple="true"></yoo-form-button-choice>
```
```yoo-form-button-choice-multiple.js

    var comp = document.querySelector('yoo-form-button-choice');
    comp.choices = ['salut', 'bonjour', 'hi', 'que tal', 'arigato', 'good morning'];

```

### Selected Attribute

```yoo-form-button-choice-selected.html

    <yoo-form-button-choice multiple="true"></yoo-form-button-choice>
```
```yoo-form-button-choice-selected.js

    var comp = document.querySelector('yoo-form-button-choice');
    comp.choices = ['salut', 'bonjour', 'hi', 'que tal', 'arigato', 'good morning'];
    comp.selected = [false, false, true, false, true, false];
    
```

## CSS

|Type|Name|Description|
|---|---|---|
|Round|`round`|Rounded label|
|Accent|`accent`|Accent style on selected|
|Danger|`danger`|Danger style on selected|
|Warning|`warning`|Warning style on selected|
|Info|`info`|Info style on selected|
|Dark|`dark`|Dark style on selected|

### Style variations

```yoo-form-button-choice-style.html

    <h3>Long choices</h3>
    <br/>
    <yoo-form-button-choice ></yoo-form-button-choice>
    <br/>
    <h3>Styles</h3>
    <br/>
    <yoo-form-button-choice class="round"></yoo-form-button-choice>
    <br/>
    <br/>    
    <yoo-form-button-choice class="accent"></yoo-form-button-choice>
    <br/>
    <br/>
    <yoo-form-button-choice class="danger"></yoo-form-button-choice>
    <br/>
    <br/>
    <yoo-form-button-choice class="warning"></yoo-form-button-choice>
    <br/>
    <br/>
    <yoo-form-button-choice class="info"></yoo-form-button-choice>
    <br/>
    <br/>
    <yoo-form-button-choice class="dark"></yoo-form-button-choice>


```
```yoo-form-button-choice-style.js

    var comps = [...document.querySelectorAll('yoo-form-button-choice')];
    comps.map((comp) => {comp.choices = ['salut', 'bonjour', 'hi', 'que tal', 'arigato', 'good morning']});
    var comp1 = document.querySelector('yoo-form-button-choice');
    comp1.choices = ['this is a verrrrrrrrrrrry looooooooooooooooooong choice', 'small', 'soooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo big choice','little'];
```

### Style variations 

## Events

|Attr|Type|Description|
|---|---|---|
|`changed`|`string[]`|When a button is clicked, an event is emitted with all the current selected values|