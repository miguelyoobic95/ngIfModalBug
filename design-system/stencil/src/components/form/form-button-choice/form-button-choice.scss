

@mixin class-color-applicator($class,$color,$color-border) {
    :host(.#{$class}) {
        .outer-container {
            .grid-container {
                .choice-container {
                    &.selected {
                        background: cssvar($color);
                        border: 1px solid cssvar($color-border);
                    }
                }
            }
        }
    }
}

:host {
    .outer-container {
        .grid-container {
            display: grid;
            grid-template-columns: 30% 30% 30%;
            grid-gap: 0.5rem;
            justify-content: center;
            .choice-container {
                background: cssvar(light);
                border: 1px solid cssvar(dark-20);
                padding: 0.1rem 0.5rem;
                text-align: center;
                color: cssvar(dark);
                @include text-truncate();
                &.selected {
                    background: cssvar(success);
                    border: 1px solid cssvar(success-110);
                    color: cssvar(light);
                }
            }
            &.few-items {
                grid-template-columns: 45% 45%;
            }
        }
    }
}

:host(.round) {
    .outer-container {
        .grid-container {
            .choice-container {
                border-radius: 5px;
            }
        }
    }
}

@include class-color-applicator(warning, warning, warning-110);
@include class-color-applicator(danger, danger, danger-110);
@include class-color-applicator(info, info, info-110);
@include class-color-applicator(accent, accent, accent-110);
@include class-color-applicator(dark, dark, dark-110);