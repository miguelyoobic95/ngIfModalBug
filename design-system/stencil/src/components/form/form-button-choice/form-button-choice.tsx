import { Component, Element, Prop, Event, EventEmitter, State } from '@stencil/core';
import { IFormSelect, ValidatorEntry, AsyncValidator, Validator, ITranslateService } from '@shared/interfaces';
import { setValidator, setAsyncValidator, setValueAndValidateInput } from '../../../utils/helpers/form-input-helpers';
import { findIndex, isEqual } from 'lodash-es';

@Component({
    tag: 'yoo-form-button-choice',
    styleUrl: 'form-button-choice.scss',
    scoped: true
})
export class YooFormButtonChoiceComponent implements IFormSelect {

    @Prop() multiple: boolean = false;
    @Prop() values: string[] = [];
    @Prop({ mutable: true }) value: Array<string> | string;
    @Prop() required: boolean;
    @Prop() validators: Array<Validator<string[]> | ValidatorEntry> = [];
    @Prop() asyncValidators: Array<AsyncValidator<string[]>>;
    @Prop() readonly: boolean;
    @Prop() useTranslate: boolean;

    @Event() validityChanged: EventEmitter<boolean>;
    @Event() inputBlurred: EventEmitter<any>;
    @Event() inputFocused: EventEmitter<boolean>;
    @Event() inputChanged: EventEmitter<any>;

    @Element() host: HTMLStencilElement;

    @State() validity: boolean;
    @State() selection: Array<string> = [];

    private translate: ITranslateService = (window as any).translateService;

    // Reduced Validators
    _validator: Validator<string> = (x: string) => true;
    _asyncValidator: AsyncValidator<string> = async (x: string) => true;

    componentWillLoad() {
        setValidator(this.validators);
        setAsyncValidator(this.asyncValidators);
    }

    componentDidLoad() {
        if (this.value) {
            this.selection = [].concat(this.value || []);
        }
    }

    hasFewItems(): boolean {
        return this.values.length < 5;
    }

    onItemSelect(item: string) {
        let index = this.selection.indexOf(item);
        if (this.multiple) {
            if (index >= 0) {
                this.selection.splice(index, 1);
                this.selection = [...this.selection];
            } else {
                this.selection = [...this.selection, item];
            }
        } else {
            if (index >= 0) {
                this.selection = [];
            } else {
                this.selection = [item];
            }
        }
        setValueAndValidateInput(this.multiple ? this.selection : (this.selection.length > 0 ? this.selection[0] : null), this);
    }

    isSelected(item) {
        let selection = [].concat(this.selection);
        let index = findIndex(selection, (s) => isEqual(item, s));
        return index >= 0;
    }

    renderItem(item: string): JSX.Element {
        return (
            <div class={'choice-container ' + (this.isSelected(item) ? 'selected' : '')} onClick={() => this.onItemSelect(item)}>
                <span>{this.useTranslate ? this.translate.get(item.toUpperCase()) : item}</span>
            </div>
        );
    }

    renderReadonly() {
        return this.value ? [].concat(this.value).map(v => <div innerHTML={v}></div>) : null;
    }

    renderEditable(): JSX.Element {
        return (
            <div class="outer-container">
                <div class={'grid-container ' + (this.hasFewItems() ? 'few-items' : '')}>
                    {this.values.map((item) => {
                        return this.renderItem(item);
                    })}
                </div>
            </div>
        );
    }

    render(): JSX.Element {
        return this.readonly ? this.renderReadonly() : this.renderEditable();
    }
}
