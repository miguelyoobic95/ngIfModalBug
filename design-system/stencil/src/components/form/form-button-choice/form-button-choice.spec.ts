import { TestWindow } from '@stencil/core/testing';
import { YooFormButtonChoiceComponent } from './form-button-choice';

describe('Form Button Choice', () => {

    it('Should build', async () => {
        let element = new YooFormButtonChoiceComponent();
        expect(element).toBeTruthy();
    });

    describe('Rendering', () => {
        it('Should render with values', async () => {
            let window = new TestWindow();
            let element = await window.load({
                components: [YooFormButtonChoiceComponent],
                html: '<yoo-form-button-choice></yoo-form-button-choice>'
              });
            element.values = ['az', 'oz', 'salut', 'bonjour', 'hi', 'max', 'min', 'chocolat'];
            await window.flush();
            expect(element.values).toEqual(['az', 'oz', 'salut', 'bonjour', 'hi', 'max', 'min', 'chocolat']);
            expect(element).toMatchSnapshot();
        });

        it('Should render with multiple prop', async () => {
            let window = new TestWindow();
            let element = await window.load({
                components: [YooFormButtonChoiceComponent],
                html: '<yoo-form-button-choice multiple="true"></yoo-form-button-choice>'
              });
            expect(element.multiple).toEqual(true);
            expect(element).toMatchSnapshot();
        });

        it('Should have few item class when values < 5', async () => {
            let window = new TestWindow();
            let element = await window.load({
                components: [YooFormButtonChoiceComponent],
                html: '<yoo-form-button-choice></yoo-form-button-choice>'
              });
            element.values = ['az', 'oz', 'salut', 'bonjour'];
            await window.flush();
            let divGrid = element.querySelector('.grid-container');
            expect(divGrid.classList).toContain('few-items');
        });

        it('Should not have few item class when values > 4', async () => {
            let window = new TestWindow();
                let element = await window.load({
                components: [YooFormButtonChoiceComponent],
                html: '<yoo-form-button-choice></yoo-form-button-choice>'
              });
            element.values = ['az', 'oz', 'salut', 'bonjour', 'saluttt'];
            await window.flush();
            let divGrid = element.querySelector('.grid-container');
            expect(divGrid.classList).not.toContain('few-items');
        });

        describe('With selected elements', () => {

            it('Should render with a pre selected element', async () => {
                let window = new TestWindow();
                let element = await window.load({
                    components: [YooFormButtonChoiceComponent],
                    html: '<yoo-form-button-choice></yoo-form-button-choice>'
                  });
                element.values = ['az', 'oz', 'salut', 'bonjour', 'saluttt'];
                element.selected = [true, false, false, false, false];
                await window.flush();
                expect(element).toMatchSnapshot();
            });

            it('Should render with to much selected elements and unselect them all', async () => {
                let window = new TestWindow();
                let element = await window.load({
                    components: [YooFormButtonChoiceComponent],
                    html: '<yoo-form-button-choice></yoo-form-button-choice>'
                  });
                element.values = ['az', 'oz', 'salut', 'bonjour', 'saluttt'];
                element.selected = [true, false, true, false, false];
                await window.flush();
                expect(element.selected.filter(val => val)).toHaveLength(0);
                expect(element).toMatchSnapshot();
            });

            it('Should render with pre selected element in multiple choice', async () => {
                let window = new TestWindow();
                let element = await window.load({
                    components: [YooFormButtonChoiceComponent],
                    html: '<yoo-form-button-choice multiple="true"></yoo-form-button-choice>'
                  });
                element.values = ['az', 'oz', 'salut', 'bonjour', 'saluttt'];
                element.selected = [true, false, true, false, true];
                await window.flush();
                expect(element).toMatchSnapshot();
            });

        });
    });

    describe('Component logic', () => {
        let element;
        beforeEach(async () => {
            element = new YooFormButtonChoiceComponent();
            element.inputChanged = {emit: jest.fn()};
        });

        it('Should create an array of the same length to store item state', async () => {
            element.values = ['a', 'b', 'c', 'd', 'e'];
            element.selectedUpdater();
            expect(element.selected).toHaveLength(5);
            expect(element.selected.filter((val) => val ? true : false)).toHaveLength(0);
        });

        it('Should only select one item if not multiple', async () => {
            element.values = ['a', 'b', 'c', 'd'];
            element.selectedUpdater();
            expect(element.selected).toEqual([false, false, false, false]);
            element.clickChoice(0);
            expect(element.selected).toEqual([true, false, false, false]);
            element.clickChoice(2);
            expect(element.selected).toEqual([false, false, true, false]);
            element.clickChoice(3);
            expect(element.selected).toEqual([false, false, false, true]);
        });

        it('Should be able to select several item if multiple is true', async () => {
            element.values = ['a', 'b', 'c', 'd'];
            element.multiple = true;
            element.selectedUpdater();
            expect(element.selected).toEqual([false, false, false, false]);
            element.clickChoice(0);
            expect(element.selected).toEqual([true, false, false, false]);
            element.clickChoice(2);
            expect(element.selected).toEqual([true, false, true, false]);
            element.clickChoice(3);
            expect(element.selected).toEqual([true, false, true, true]);
            element.clickChoice(0);
            expect(element.selected).toEqual([false, false, true, true]);
            element.clickChoice(3);
            expect(element.selected).toEqual([false, false, true, false]);
            element.clickChoice(1);
            expect(element.selected).toEqual([false, true, true, false]);
        });
    });

  describe('events', () => {
      let window, element, res, emitEvent;
      beforeEach( async () => {
        window = new TestWindow();
        element = await window.load({
          components: [YooFormButtonChoiceComponent],
            html: '<yoo-form-button-choice></yoo-form-button-choice>'
        });
        element.values = ['az', 'oz', 'salut', 'bonjour', 'hi', 'max', 'min', 'chocolat'];
        await window.flush();
        emitEvent = jest.fn(ev => { res = ev.detail; });
      });


    it ('should emit changed on choice-container click', async () => {
        window.document.addEventListener('inputChanged', emitEvent);
        let button = element.querySelector('.choice-container');
        await button.click();
        expect(emitEvent).toHaveBeenCalled();
        expect(res).toEqual([element.values[0]]);
    });
  });
});
