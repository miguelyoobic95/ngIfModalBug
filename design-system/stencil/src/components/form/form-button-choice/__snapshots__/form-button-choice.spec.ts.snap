// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Button Choice Rendering Should render with multiple prop 1`] = `
<yoo-form-button-choice
  class="hydrated"
  data-yoo-form-button-choice-host=""
  multiple="true"
>
  
  <div
    class="outer-container"
    data-yoo-form-button-choice=""
  >
    <div
      class="grid-container few-items"
      data-yoo-form-button-choice=""
    />
  </div>
</yoo-form-button-choice>
`;

exports[`Form Button Choice Rendering Should render with values 1`] = `
<yoo-form-button-choice
  class="hydrated"
  data-yoo-form-button-choice-host=""
>
  
  <div
    class="outer-container"
    data-yoo-form-button-choice=""
  >
    <div
      class="grid-container"
      data-yoo-form-button-choice=""
    >
      <div
        class="choice-container"
        data-yoo-form-button-choice=""
      >
        <span
          data-yoo-form-button-choice=""
        >
          az
        </span>
      </div>
      <div
        class="choice-container"
        data-yoo-form-button-choice=""
      >
        <span
          data-yoo-form-button-choice=""
        >
          oz
        </span>
      </div>
      <div
        class="choice-container"
        data-yoo-form-button-choice=""
      >
        <span
          data-yoo-form-button-choice=""
        >
          salut
        </span>
      </div>
      <div
        class="choice-container"
        data-yoo-form-button-choice=""
      >
        <span
          data-yoo-form-button-choice=""
        >
          bonjour
        </span>
      </div>
      <div
        class="choice-container"
        data-yoo-form-button-choice=""
      >
        <span
          data-yoo-form-button-choice=""
        >
          hi
        </span>
      </div>
      <div
        class="choice-container"
        data-yoo-form-button-choice=""
      >
        <span
          data-yoo-form-button-choice=""
        >
          max
        </span>
      </div>
      <div
        class="choice-container"
        data-yoo-form-button-choice=""
      >
        <span
          data-yoo-form-button-choice=""
        >
          min
        </span>
      </div>
      <div
        class="choice-container"
        data-yoo-form-button-choice=""
      >
        <span
          data-yoo-form-button-choice=""
        >
          chocolat
        </span>
      </div>
    </div>
  </div>
</yoo-form-button-choice>
`;

exports[`Form Button Choice Rendering With selected elements Should render with a pre selected element 1`] = `
<yoo-form-button-choice
  class="hydrated"
  data-yoo-form-button-choice-host=""
>
  
  <div
    class="outer-container"
    data-yoo-form-button-choice=""
  >
    <div
      class="grid-container"
      data-yoo-form-button-choice=""
    >
      <div
        class="choice-container"
        data-yoo-form-button-choice=""
      >
        <span
          data-yoo-form-button-choice=""
        >
          az
        </span>
      </div>
      <div
        class="choice-container"
        data-yoo-form-button-choice=""
      >
        <span
          data-yoo-form-button-choice=""
        >
          oz
        </span>
      </div>
      <div
        class="choice-container"
        data-yoo-form-button-choice=""
      >
        <span
          data-yoo-form-button-choice=""
        >
          salut
        </span>
      </div>
      <div
        class="choice-container"
        data-yoo-form-button-choice=""
      >
        <span
          data-yoo-form-button-choice=""
        >
          bonjour
        </span>
      </div>
      <div
        class="choice-container"
        data-yoo-form-button-choice=""
      >
        <span
          data-yoo-form-button-choice=""
        >
          saluttt
        </span>
      </div>
    </div>
  </div>
</yoo-form-button-choice>
`;

exports[`Form Button Choice Rendering With selected elements Should render with pre selected element in multiple choice 1`] = `
<yoo-form-button-choice
  class="hydrated"
  data-yoo-form-button-choice-host=""
  multiple="true"
>
  
  <div
    class="outer-container"
    data-yoo-form-button-choice=""
  >
    <div
      class="grid-container"
      data-yoo-form-button-choice=""
    >
      <div
        class="choice-container"
        data-yoo-form-button-choice=""
      >
        <span
          data-yoo-form-button-choice=""
        >
          az
        </span>
      </div>
      <div
        class="choice-container"
        data-yoo-form-button-choice=""
      >
        <span
          data-yoo-form-button-choice=""
        >
          oz
        </span>
      </div>
      <div
        class="choice-container"
        data-yoo-form-button-choice=""
      >
        <span
          data-yoo-form-button-choice=""
        >
          salut
        </span>
      </div>
      <div
        class="choice-container"
        data-yoo-form-button-choice=""
      >
        <span
          data-yoo-form-button-choice=""
        >
          bonjour
        </span>
      </div>
      <div
        class="choice-container"
        data-yoo-form-button-choice=""
      >
        <span
          data-yoo-form-button-choice=""
        >
          saluttt
        </span>
      </div>
    </div>
  </div>
</yoo-form-button-choice>
`;

exports[`Form Button Choice Rendering With selected elements Should render with to much selected elements and unselect them all 1`] = `
<yoo-form-button-choice
  class="hydrated"
  data-yoo-form-button-choice-host=""
>
  
  <div
    class="outer-container"
    data-yoo-form-button-choice=""
  >
    <div
      class="grid-container"
      data-yoo-form-button-choice=""
    >
      <div
        class="choice-container"
        data-yoo-form-button-choice=""
      >
        <span
          data-yoo-form-button-choice=""
        >
          az
        </span>
      </div>
      <div
        class="choice-container"
        data-yoo-form-button-choice=""
      >
        <span
          data-yoo-form-button-choice=""
        >
          oz
        </span>
      </div>
      <div
        class="choice-container"
        data-yoo-form-button-choice=""
      >
        <span
          data-yoo-form-button-choice=""
        >
          salut
        </span>
      </div>
      <div
        class="choice-container"
        data-yoo-form-button-choice=""
      >
        <span
          data-yoo-form-button-choice=""
        >
          bonjour
        </span>
      </div>
      <div
        class="choice-container"
        data-yoo-form-button-choice=""
      >
        <span
          data-yoo-form-button-choice=""
        >
          saluttt
        </span>
      </div>
    </div>
  </div>
</yoo-form-button-choice>
`;
